// @stitches/react/types/typings/Issue-697-css.tsx
import React from 'react'
import * as Stitches from '../index.js'

const styled = Stitches.css('span', {
	color: 'red',
	variants: {
		/** Size variant */
		mySize: {
			myLarge: {},
		},
	},
})

const Styled = function Styled(
	props: Stitches.ComponentProps<typeof styled>
) {
	return (
		<div {...styled({
			mySize: 'myLarge',
			css: {
				color: 'ActiveText',
			},
		}).props}>button</div>
	)
}

export function StyledPage() {
	return (
		<Styled mySize="myLarge" css={{
			color: 'ActiveText'
		}}>button</Styled>
	)
}

const StyledFunction = function StyledFunction(
	props: React.ComponentProps<typeof Styled>
) {
	return null
}

export function StyledFunctionPage() {
	return (
		<StyledFunction
			mySize="myLarge"
			css={{
				color: 'ActiveText'
			}}
		>button</StyledFunction>
	)
}

const StyledFunction2 = function StyledFunction(
	props: React.ComponentProps<typeof Styled>
) {
	return null
}

export function StyledFunction2Page() {
	return (
		<StyledFunction2
			mySize="myLarge"
			as="a" // not supported, intentional, StyledFunction2 is a regular function
			css={{
				color: 'ActiveText'
			}}
		>button</StyledFunction2>
	)
}

export const StyledForwardRef: React.ForwardRefExoticComponent<React.ComponentProps<typeof Styled>> = React.forwardRef(
	(props, forwardedRef) => {
		return <StyledFunction {...props} />;
	}
)

export default function StyledForwardRefPage() {
	return <StyledForwardRef />
}
