// @stitches/react/types/typings/Issue-697.tsx
import React from 'react'
import * as Stitches from '../index.js'

const Styled = Stitches.styled('span', {
	color: 'red',
	variants: {
		/** The **mySize** variant controls the size of the element. */
		mySize: {
			myLarge: {},
		},
	},
})

export function StyledPage() {
	return (
		<Styled mySize="myLarge" css={{
			color: 'DarkBlue',
		}}>button</Styled>
	)
}

const StyledFunction = function StyledFunction(
	props: React.ComponentProps<typeof Styled>
) {
	return null
}

export function StyledFunctionPage() {
	return (
		<StyledFunction
			mySize="myLarge"
			css={{
				color: 'DarkBlue',
			}}
		>button</StyledFunction>
	)
}

const StyledFunction2 = function StyledFunction(
	props: React.ComponentProps<typeof Styled>
) {
	return null
}

export function StyledFunction2Page() {
	return (
		<StyledFunction2
			mySize="myLarge"
			css={{
				color: 'ActiveText',
			}}
		>button</StyledFunction2>
	)
}

export const StyledForwardRef = React.forwardRef<
	HTMLSpanElement,
	React.ComponentProps<typeof Styled>
>(
	(props, forwardedRef) => {
		return <StyledFunction {...props} />;
	}
)

export function StyledForwardRefPage() {
	return <StyledForwardRef mySize="myLarge" css={{
		color: 'ActiveText'
	}} />
}

const Generic = function(props: { happy?: true }) {
	return null
}

export const GenericForwardRef = React.forwardRef<
	HTMLSpanElement,
	React.ComponentProps<typeof Generic>
>(
	(props, forwardedRef) => {
		return <Generic {...props} />;
	}
)

export function GenericForwardRefPage() {
	return <StyledForwardRef mySize="myLarge" css={{
		color: 'ActiveText',
	}} />
}
